src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/interior_data_manager_impl.cc

```diff
@@ -0,0 +1,123 @@
+#include "rc_rpc_plugin/interior_data_manager_impl.h"
+#include "rc_rpc_plugin/rc_helpers.h"
+#include "rc_rpc_plugin/rc_rpc_plugin.h"
+#include "application_manager/application_manager.h"
+#include "application_manager/rpc_service.h"
+
+namespace rc_rpc_plugin {
+CREATE_LOGGERPTR_GLOBAL(logger_, "RemoteControlModule");
+
+InteriorDataManagerImpl::InteriorDataManagerImpl(
+    RCRPCPlugin& rc_plugin,
+    InteriorDataCache& cache,
+    application_manager::ApplicationManager& app_mngr,
+    application_manager::rpc_service::RPCService& rpc_service)
+    : rc_plugin_(rc_plugin)
+    , cache_(cache)
+    , app_mngr_(app_mngr)
+    , rpc_service_(rpc_service) {}
+
+void InteriorDataManagerImpl::OnPolicyEvent(plugins::PolicyEvent event) {
+  UpdateHMISubscriptionsOnPolicyUpdated();
+}
+
+void InteriorDataManagerImpl::OnApplicationEvent(
+    plugins::ApplicationEvent event,
+    app_mngr::ApplicationSharedPtr application) {
+  if (plugins::ApplicationEvent::kApplicationUnregistered == event ||
+      plugins::ApplicationEvent::kApplicationExit == event) {
+    UpdateHMISubscriptionsOnAppUnregistered(*application);
+  }
+}
+
+void InteriorDataManagerImpl::OnDisablingRC() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  auto existing_subscription = AppsSubscribedModules();
+  std::set<std::string> subscribed_modules;
+  for (auto& pair : existing_subscription) {
+    auto& app = pair.first;
+    auto rc_extension = RCHelpers::GetRCExtension(*app);
+    for (const auto& module : pair.second) {
+      subscribed_modules.insert(module);
+      rc_extension->UnsubscribeFromInteriorVehicleData(module);
+    }
+  }
+  for (auto& module : subscribed_modules) {
+    LOG4CXX_TRACE(logger_, "unsubscribe " << module);
+    UnsubscribeFromInteriorVehicleData(module);
+  }
+}
+
+void InteriorDataManagerImpl::UpdateHMISubscriptionsOnPolicyUpdated() {
+  auto apps_allowed_modules =
+      RCHelpers::GetApplicaitonsAllowedModules(app_mngr_);
+  auto apps_subscribed_modules = AppsSubscribedModules();
+  InteriorDataManagerImpl::AppsModules apps_disallowed_modules;
+  for (auto& pair : apps_subscribed_modules) {
+    auto& allowed = apps_allowed_modules[pair.first];
+    auto& subscribed = pair.second;
+    std::vector<std::string> disallowed_modules;
+    std::set_difference(subscribed.begin(),
+                        subscribed.end(),
+                        allowed.begin(),
+                        allowed.end(),
+                        std::back_inserter(disallowed_modules));
+    apps_disallowed_modules[pair.first] = disallowed_modules;
+  }
+
+  for (auto& pair : apps_disallowed_modules) {
+    auto& app = pair.first;
+    auto rc_extension = RCHelpers::GetRCExtension(*app);
+    for (const auto& module : pair.second) {
+      rc_extension->UnsubscribeFromInteriorVehicleData(module);
+      auto apps_subscribed =
+          RCHelpers::AppsSubscribedToModuleType(module, app_mngr_);
```

Comment:
does this "apps_subscribed = RCHelpers::AppsSubscribedToModuleType(module, app_mngr_);" count the current app?

Reply:
@yang1070 it counts all registered applications but in the line above we unsubscribed current app from module type, so it won't be in this map with current module type but can be with some other module types